pyam/_aggregate.py

```diff
@@ -42,76 +52,62 @@ def _aggregate(df, variable, components=None, method=np.sum):
+    data_list = []
 
-    # iterate over subcategories (bottom-up) and perform aggregation
-    for entry in sub_variables:
-        _df.aggregate(variable=entry, append=True)
-        _df_temp = _df.aggregate(variable=entry, append=False)
+    # iterate over variables (bottom-up) and aggregate all components up to `variable`
+    for d in reversed(range(find_depth(variable), max(find_depth(_df.variable)))):
+        components = compress(_df.variable, find_depth(_df.variable, level=d + 1))
+        var_list = set([reduce_hierarchy(v, -1) for v in components])
 
-        if _df_aggregated is None:
-            _df_aggregated = _df_temp.copy()
-        else:
-            _df_aggregated.append(_df_temp, inplace=True)
+        # a temporary dataframe allows to distinguish between full data and new data
+        temp_df = _df.aggregate(variable=var_list)
+        _df.append(temp_df, inplace=True)
+        data_list.append(temp_df._data)
 
-    return _df_aggregated.data
+    return pd.concat(data_list)
 
 
 def _aggregate_region(
     df, variable, region, subregions=None, components=False, method="sum", weight=None
 ):
     """Internal implementation for aggregating data over subregions"""
     if not isstr(variable) and components is not False:
-        msg = "aggregating by list of variables with components " "is not supported"
-        raise ValueError(msg)
+        raise ValueError(
+            "Aggregating by list of variables with components is not supported!"
+        )
 
     if weight is not None and components is not False:
-        msg = "using weights and components in one operation not supported"
-        raise ValueError(msg)
+        raise ValueError("Using weights and components in one operation not supported!")
 
     # default subregions to all regions other than `region`
     subregions = subregions or df._all_other_regions(region, variable)
 
     if not len(subregions):
-        msg = (
-            "cannot aggregate variable `{}` to `{}` because it does not"
-            " exist in any subregion"
+        logger.info(
+            f"Cannot aggregate variable '{variable}' to '{region}' "
+            "because it does not exist in any subregion!"
         )
-        logger.info(msg.format(variable, region))
-
         return
 
     # compute aggregate over all subregions
     subregion_df = df.filter(region=subregions)
     rows = subregion_df._apply_filters(variable=variable)
     if weight is None:
-        col = "region"
-        _data = _group_and_agg(subregion_df.data[rows], col, method=method)
+        _data = _group_and_agg(subregion_df._data[rows], "region", method=method)
     else:
         weight_rows = subregion_df._apply_filters(variable=weight)
         _data = _agg_weight(
-            subregion_df.data[rows], subregion_df.data[weight_rows], method
+            subregion_df._data[rows], subregion_df._data[weight_rows], method
         )
 
     # if not `components=False`, add components at the `region` level
```

Comment:
I'm unsure what the correct behaviour for `components=[]` would be, here it is treated the same as `components=False`. If that is correct i'd suggest using `if components:` instead of the `if components is not False`.

Reply:
In the untouched lines below this snippet :/

https://github.com/IAMconsortium/pyam/blob/8ca67945d55928cbc615b624a51dfec20079acd4/pyam/_aggregate.py#L114-L132

Reply:
Agree, if you explicitly pass an empty list of components, that should be equivalent to setting components to False...

Reply:
Changed as suggested